### ISSUE TITLE
New coupling scenarios

### ISSUE THREAD

**Issue by ktht** [HEP-KBFI/hh-multilepton#38]
Adding `8a` and another 7 BMs. See [here](https://gitlab.cern.ch/hh/eft-benchmarks).

**Comment by tolange** [comment]
Should we add box as well? (kl=cg=c2g=c2=0, kt=1)

**Comment by tolange** [comment]
![BM2018_8a](https://user-images.githubusercontent.com/11850680/115050871-8073d280-9edc-11eb-8be6-9c929cc1c583.png)
![BM2020_0](https://user-images.githubusercontent.com/11850680/115050877-810c6900-9edc-11eb-85e4-6cf306fc582b.png)
![BM2020_1](https://user-images.githubusercontent.com/11850680/115050879-81a4ff80-9edc-11eb-8804-88e45613e025.png)
![BM2020_2](https://user-images.githubusercontent.com/11850680/115050880-81a4ff80-9edc-11eb-964c-4f77cae6cbb4.png)
![BM2020_3](https://user-images.githubusercontent.com/11850680/115050882-823d9600-9edc-11eb-8511-0e758522f439.png)
![BM2020_4](https://user-images.githubusercontent.com/11850680/115050883-82d62c80-9edc-11eb-9963-13a45cc08653.png)
![BM2020_5](https://user-images.githubusercontent.com/11850680/115050885-82d62c80-9edc-11eb-8cb6-ffdf07df4b1f.png)
![BM2020_6](https://user-images.githubusercontent.com/11850680/115050886-82d62c80-9edc-11eb-9790-eec3009b1a69.png)
Mapped the closest BM shapes between the old and new scenarios to find the optimal BDT output minimizing the shape difference formula also used in combine: https://cms-analysis.github.io/HiggsAnalysis-CombinedLimit/part3/validation/#uncertainty-probably-has-no-genuine-shape-effect i.e. Sum (2*|h1(i)-h2(i)|/|h1(i)+h2(i)|). We get (new->old): JHEP03BM1-> JHEP04BM11, JHEP03BM2->JHEP04BM11, JHEP03BM3->JHEP04BM5, JHEP03BM4->JHEP04BM3, JHEP03BM5->JHEP04BM9, JHEP03BM6->JHEP04BM3, JHEP03BM7->JHEP04BM9, 8a->JHEP04BM9

**Comment by ktht** [comment]
@tolange Refactored the LO normalization part of the HH reweighting as promised. I added these lines
```cpp
{
  const HHWeightInterfaceLO HHWeightLO_calc(&couplings, hhWeight);
  const std::vector<double> norm = HHWeightLO_calc.getNorm();
  std::cout << "LO normalization coefficients: " << std::fixed << std::setprecision(5) << format_vdouble(norm) << '\n';
}
```
to [`testHHWeightInterfaceNLO.cc`](https://github.com/HEP-KBFI/tth-htt/blob/master/bin/testHHWeightInterfaceNLO.cc) to make sure that the new code reproduces the hardcoded set of normalization coefficients (([ink](https://github.com/HEP-KBFI/tth-htt/blob/3690e640ce2b6ae694c34dfdd28b3f0e5981bee3/src/HHWeightInterfaceCouplings.cc#L33)) and it all seems to check out. However, after introducing `HHWeightInterfaceLO` to the code I now get this error:
```
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
ImportError: No module named site
```
I'm not sure why this error doesn't appear in the normal analysis, and why it even complains about the python variables. I guess maybe it has something to do with the grid setup that could potentially spoil some environment variables. It is possible to get around this error by resetting `$PYTHONPATH` to CMSSW default for the runtime (not the backslash at the end -> run as a single line):
```bash
PYTHONHOME=$(cd $CMSSW_BASE; scram tool info python | grep PYTHON_BASE | tr '=' ' ' | awk '{print $NF}') \
testHHWeightInterfaceNLO
```
The reason why I'm mentioning all this is because the test code now also includes an example of how to query reweighting weights for the c2 scan. So it's just a matter of computing a weighted average of the reweighting weights over the cosTheta star bins in every mHH bin.

**Comment by tolange** [comment]
Tested the possibility to create the 56 c2 points needed by the linear combination of 3 delicately choosen points in c2. c2=0.1 and c2=0.35 seem to be the most extreme cases in the scan going from -2 to 3. I tried to replicate Lucas [method](https://login.cern.ch/adfs/ls/?wtrealm=https%3a%2f%2foauth.web.cern.ch%2f&wctx=WsFedOwinState%3d9XTFqt2MNWqsKoUoEb_tQ_I-m6vaYQnDWTYRHjNIfYgLJhmtHuCm4dWuuUlHrKkF6IhAKJJujEYvkDgvT4jOOr6OdMWRrt1bR7ZgXQFogwZVYB8MKdg7W93qsJMKpIeP0FYSOcSKleKgzoQbCi_-Jk6-aaY3a30YoFKoN6mqEblum1AmQKXp-Ispjhh224L5bDRIWv8s23ofntjK7BU4ZRrwEdb8PbHkc7rBi0xQNZ6kjUHqvQ7jqsk_ywJDEz1AwY9zCZGvMe9Gh923j7q4AnuD4rGsCmD5kRwdENkzYjD4nER8&wa=wsignin1.0) (around slide 5). I end up with a parametrization in c2, c2^2 and a constant i.e. c={1,c2^2,c2} choosing SM, 0.1 and 0.35 I get C={{1,0,0},{1,0.01,0.1},{1,0.35^2,0.35}} and thus with C^-1={{1,0,0},{28.5714,-40,11.4286},{-12,8571,14,-1.14286}}. With this I can reproduce any point in the grid sufficiently well provided the input histograms are scale to the proper cross section. Regarding this issue this means we can get away with 2 additional points (c2=0.1,0.35) and maybe 2 more for validation on reco level. I would suggest the end points of the grid -2 and 3. More information on Friday. I will write a tool to create the 56 datacards this way out of the 3 c2 points.
[c2scan.pdf](https://github.com/HEP-KBFI/hh-multilepton/files/6345342/c2scan.pdf)
[c_c2_3p00.pdf](https://github.com/HEP-KBFI/hh-multilepton/files/6345343/c_c2_3p00.pdf)
[c_c2_m2p00.pdf](https://github.com/HEP-KBFI/hh-multilepton/files/6345344/c_c2_m2p00.pdf)
[c_c2_0p10.pdf](https://github.com/HEP-KBFI/hh-multilepton/files/6345345/c_c2_0p10.pdf)
[c_c2_0p00.pdf](https://github.com/HEP-KBFI/hh-multilepton/files/6345346/c_c2_0p00.pdf)
[c_c2_0p35.pdf](https://github.com/HEP-KBFI/hh-multilepton/files/6345347/c_c2_0p35.pdf)

**Comment by acarvalh** [comment]
super!

Better further scripting however is to add the parametrization in another physics model, eg HHModelPinv_c2.py (just following the one in inference [here](https://gitlab.cern.ch/hh/tools/inference/-/blob/master/dhi/models/HHModelPinv.py) == just substitute the parameterization for GGF), then you will not need to  I will write a tool to create the 56 datacards, but one card with several signals as you do for kl-kt scans.
+ that becomes a group contribution, and nobody else will need to do 56 points

**Comment by tolange** [comment]
I agree that would be better, the idea of writing a tool for 56 datacards is just out there as this is currently the proposal for the combination. But that might be an discussion for the next HH meeting on tuesday. If other analysis are on board, we also can do this with a physics model and save some computation (and human) resources :)

**Comment by acarvalh** [comment]
It was an idea with me and conveners since some time ;-) I could not get to that yet. It will be very welcome.
Bonus: we could make k x c2 correlation.

With a validation, I am sure everybody will be onboard. 
You can present next meeting even if only the validation of the combinationn wrt selected points if you cannot get to physics model yet, so we can converge, people can avoid doing the 56 cards and implement that limit paralelization in inference -- but that should be eazy as there is the template.

**Comment by tolange** [comment]
I remember ;) But I actually didnt use actual simulation yet, I used the Theo cross sections as they come out of the reweighing. But if it holds up there, before things enter the analysis, it should also hold up for the final analysis outputs. Ok, for a 2D/3D scan it gets slightly more complicated, there we need 6 histograms in total (similar to vbf) the parametrization would then be in kt^2*kl^2, kt^4, c2^2, c2*kt*kl, c2*kt^2, kt^3*kl. As the points chosen need to be able to form a linear independent basis we need points away from the c2 axis in the c2-kt and c2-kl plane as well. As we already use them for the kl-kt scans I thought of kl=0,5,2.45 but it seems that is not sufficient, we need 2 of those and one more like (kl=0,kt=1,c2=1). So I choose the three from the only c2 scan SM, c2=0.1, c2=0.35, kl=0 (box like), kl=5 (triangle like) and the new one (kl=0,kt=1,c2=1) (>< + [], so [x] if we need a symbol/name) and this seems to work at least for the c2 scan, kt/kl I would need to verify but I see no reason why it should not work. I will prepare a few slides for Tuesdays meeting then :)

**Comment by ktht** [comment]
Quick preview of how the output of analysis jobs is going to look like:
![Screenshot_20210426_151144](https://user-images.githubusercontent.com/6233872/116080544-df3a0880-a6a1-11eb-96f1-a2839aaf066c.png)
The mapping of new BMs to existing trainings is already done -- I need to just commit the changes.

It still requires extensive testing to make sure that the changes don't break anything. Also, bbWW needs some care as well, although the focus has shifted towards resonant analysis for us.

**Comment by ktht** [comment]
So I ran 2lss workflow with acceptance uncertainties (`-s internal` option) on the master branch and on the development branch, and compared the results. I get exactly the same histograms back for the matching benchmarks (eg `BM1` and `JHEP04BM1`), which confirms that the refactoring didn't have any effect on previously established BM scenarios.

The main changes are:
- BMs listed [here](https://gitlab.cern.ch/hh/eft-benchmarks) are hardcoded in `HHWeightInterfaceCouplings` class where the association b/w the BM and the training is also created, and in `hhSettings.py` hosted in the ttH repository;
- the scan points, such as for the c2 scan, are defined with a text file hosted in this repository:
https://github.com/HEP-KBFI/hh-multilepton/blob/f18243242364cdffc55a5d2aaf3d73e99cb37dde/data/c2_scan.dat#L1-L4
- which BMs to reweight to is basically controlled by which histograms the user asks the datacards for;
- the functions that book MVAs are split into resonant and non-resonant cases (eg `CreateBDTOutputMap` is split into `CreateResonantBDTOutputMap` and `CreateResonantBDTOutputMap`).

Regarding the memory consumption, I don't think it's going to be an issue in HH-multilepton. We should be able to afford two more scan points that we can use to demonstrate that interpolation between fixed c2 scan points works.

In HH-bbww, when I tested DL interactively, to my surprise the memory consumption stayed below 1 GB for a job that also ran with acceptance uncertainties. Maybe I tested on too few events, although usually the memory usage blows up with histogram creation.

I'll now merge the development branches into the master branches.
